src/mame/drivers/ymdx100.cpp

```diff
@@ -92,20 +283,258 @@ void yamaha_dx100_state::mem_map(address_map &map)
+
+	PORT_START("AN1")
+	// The mod wheel stays in place to wherever it's set.
+	PORT_BIT( 0xFF, 0, IPT_POSITIONAL ) PORT_NAME("Modulation Wheel") PORT_SENSITIVITY(100) PORT_KEYDELTA(10) PORT_MINMAX(0x00, 0xFF) PORT_CENTERDELTA(0)
+
+	PORT_START("AN2")
+	// TODO I have no idea what kind of input this should actually be...
+	// TODO also this appears to be inverted; if this is set to 255 it behaves as if
+	// there was no breath controller? or at least seems to? on instruments like
+	// 112 Pianobrass if this is set to 0 it acts as if the mod wheel had been turned
+	// all the way up and enables LFO -- and there's probably a better way we could
+	// simulate not having a breath controller attached at all???
+	PORT_BIT( 0xFF, 0, IPT_POSITIONAL ) PORT_NAME("Breath Controller") PORT_SENSITIVITY(100) PORT_KEYDELTA(10) PORT_MINMAX(0x00, 0xFF) PORT_CENTERDELTA(0) PORT_REVERSE
+
+	PORT_START("AN3")
+	// The data entry slider stays in place to wherever it's set.
+	PORT_BIT( 0xFF, 0, IPT_POSITIONAL ) PORT_NAME("Data Entry Slider") PORT_SENSITIVITY(100) PORT_KEYDELTA(10) PORT_MINMAX(0x00, 0xFF) PORT_CENTERDELTA(0)
+
+	PORT_START("AN5")
+	PORT_CONFNAME( 0xFF, 0x80, "Internal RAM Battery Level" )
+	// "CNG RAM BATTERY!" displayed unless value is between 0x70 and 0xCC
+	PORT_CONFSETTING( 0x6F, "Too Low" )
+	PORT_CONFSETTING( 0x70, "Lowest Allowed" )
+	PORT_CONFSETTING( 0x80, "Normal" ) // for some arbitrary definition of "normal"
+	PORT_CONFSETTING( 0xCB, "Highest Allowed" )
+	PORT_CONFSETTING( 0xCC, "Too High" )
+
+	PORT_START("AN6")
+	PORT_CONFNAME( 0xFF, 0x00, "Battery Power Level" )
+	// Yes, higher values mean lower voltages.
+	// I think this is opposite to how the RAM battery voltmeter works.
+	// The weird granularity here is due to the buggy implementation of the
+	// test in the test mode; all 7V values should LED flash, but some don't.
+	PORT_CONFSETTING( 0x00, "9V (Normal)" ) // for some arbitrary definition of "normal"
+	PORT_CONFSETTING( 0x4B, "9V (Lowest Allowed)" )
+	PORT_CONFSETTING( 0x67, "7V (Highest Allowed Without LED Flash)" )
+	PORT_CONFSETTING( 0x6B, "7V (Lowest Allowed Without LED Flash)" )
+	PORT_CONFSETTING( 0x6C, "7V (Highest Allowed With LED Flash)" )
+	PORT_CONFSETTING( 0x6F, "7V (Lowest Allowed With LED Flash)" )
+	PORT_CONFSETTING( 0x70, "Less Than 7V" )
 INPUT_PORTS_END
 
 void yamaha_dx100_state::dx100(machine_config &config)
 {
 	HD6303X(config, m_maincpu, 7.15909_MHz_XTAL / 2); // HD6303XP
 	m_maincpu->set_addrmap(AS_PROGRAM, &yamaha_dx100_state::mem_map);
-	m_maincpu->in_p2_cb().set([this]() -> u8 { return m_midi_in ? 0x04 : 0; });
-	m_maincpu->in_p6_cb().set("adc", FUNC(m58990_device::eoc_r)).lshift(4);
+	m_maincpu->in_p2_cb().set_ioport("P2");
+	m_maincpu->out_p2_cb().set_ioport("P2");
+	m_maincpu->in_p6_cb().set([this]() -> u8 {
+		u8 casplay = m_port6_val & 0x80; // [TODO1]
+		u8 casrec = m_port6_val & 0x40; // [TODO1]
+		u8 adcval = ((u8) (m_adc->eoc_r() << 4)) & 0x10;
+		u8 port5line = (m_port6_val & 0x2F);
```

Comment:
Hex notations are case inconsistent across the file, they should be lowercase for all.

Reply:
Done.